Moved the metadata into `setup.cfg`

KOLANICH opened the pull request:
Added `pyproject.toml`.
Version is now populated automatically from git tags using `setuptools_scm`. Deleted `bumpversion.sh`.
Fixed some missing metadata fields.

JeffLIrion commented:
I've never used a `setup.cfg` file before, what is the benefit over using `setup.py`?

Also, the CI checks did not pass.

KOLANICH commented:
>I've never used a setup.cfg file before, what is the benefit over using setup.py?

No strong benefits currently:
* it is still possible to organize package vendor-supplied code execution while building a package with `setup.cfg` + `pyproject.toml`, but it was discussed that the long-term goal is to eliminate package vendor-supplied code execution in this case
* even more, soon `setup.cfg` will deprecated and its contents should be recommended to be moved into `pyproject.toml` (currently there is no automatic convertor `setup.cfg` -> `pyproject.toml`, but it will be more automatic, more secure and more reliable because of more constrained and parseable structure of `setup.cfg`, than `setup.py` -> `setup.cfg`, I guess even a bot automatically sending PRs will be possible).

JeffLIrion commented:
From some brief reading, it looks like `pyproject.toml` is the way of the future. So if you get the CI checks to pass and once I understand how it works, I'll merge this.

JeffLIrion commented on .github/workflows/python-package.yml:
```diff
@@ -25,10 +25,13 @@ jobs:
+        pip install ./dist/*.whl;
         if python --version 2>&1 | grep -q "Python 3.7" || python --version 2>&1 | grep -q "Python 3.8" || python --version 2>&1 | grep -q "Python 3.9" || python --version 2>&1 | grep -q "Python 3.10"; then pip install aiofiles; fi
     - name: Lint with pylint and flake8
       run: |
```
I think you need to exclude `version.py` from the linting checks below.

JeffLIrion replied:
Nevermind, it looks like you added a module docstring to this file.

JeffLIrion commented on setup.py:
```diff
@@ -1,24 +0,0 @@
-from setuptools import setup
```
It might be worth keeping a shim version of this file. I'm not sure. 

```python
import setuptools

if __name__ == "__main__":
    setuptools.setup()
```

https://stackoverflow.com/a/62983901

KOLANICH replied:
>If you need a `setup.py` for an editable install

You don't need it thanks to my PR into pip ;) You onpy need the latest `pip`.

JeffLIrion commented on .gitignore:
```diff
@@ -1,3 +1,6 @@
+# Autogenerated version file
+/adb_shell/version.py
```
If a user doesn't have this file, then they won't be able to test the package locally. Is there a command to generate `version.py`?

KOLANICH replied:
`python3 -m build -nwx` It builds a wheel, and `setuptools_scm` generates the file in process.

JeffLIrion commented on .github/workflows/python-package.yml:
```diff
@@ -41,3 +43,8 @@ jobs:
       run: |
         if python --version 2>&1 | grep -q "Python 2" || python --version 2>&1 | grep -q "Python 3.5" || python --version 2>&1 | grep -q "Python 3.6" ; then for synctest in $(cd tests && ls test*.py | grep -v async); do python -m unittest discover -s tests/ -t . -p "$synctest" || exit 1; done; fi
         if python --version 2>&1 | grep -q "Python 3.7" || python --version 2>&1 | grep -q "Python 3.8" || python --version 2>&1 | grep -q "Python 3.9" || python --version 2>&1 | grep -q "Python 3.10"; then coverage run --source adb_shell -m unittest discover -s tests/ -t . && coverage report -m && coveralls; fi
+    - name: Upload wheel as a workflow artifact
```
I haven't used GitHub Actions to upload an artifact. What's the benefit of doing so?

arichr replied:
> What's the benefit of doing so?

The ability to download created files (e.g. wheels) to use them somehow. For example, to get a proper wheel before making a release.

JeffLIrion commented on setup.cfg:
```diff
@@ -0,0 +1,31 @@
+keywords = adb, android
+url = https://github.com/JeffLIrion/adb_shell
+long_description = file: README.rst
+long_description_contet_type = text/x-rst
```
```suggestion
long_description_content_type = text/x-rst
```

JeffLIrion commented on Makefile:
```diff
@@ -3,7 +3,7 @@ release:
 	rm -rf build
 	scripts/git_tag.sh
-	python setup.py sdist bdist_wheel
+	python -m build -nwsx
```
I don't think this will work... What this target did was get the version by parsing `__init__.py`, create a git tag of the form `v{version}` (e.g., `v0.4.0`), then build the package and upload it to PyPi. I would call this after running `bumpversion.sh` to update the package version in the code. 

I think your workflow is the opposite: first create a git tag, then use that tag as the package version when building and uploading the package.

KOLANICH commented:
1. I have fixed the scripts to take the version from git tags and to update it. It will only work if there are commits upon the previous tag. Otherwise one should use the `retag` script.
2. I have cleaned up the scripts a bit. Some used to be determining package name and then discarding it. That code has been removed.

>I haven't used GitHub Actions to upload an artifact. What's the benefit of doing so?

Not really a strong benefit, but sometimes allows you to get the artifact build on the CI. Can have various uses, from not-building the package as a part of other CI pipelines and using the wheel, to troubleshooting the issues on CI.
